_posts/2019-08-08-hacking-for-cruise-control.md

```diff
@@ -0,0 +1,298 @@
+```yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    image: <registry>/<kafka-image-name>
+    version: 2.3.0
+    replicas: 3
+    listeners:
+      plain: {}
+      tls: {}
+    config:
+      metric.reporters: "com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter"
+      offsets.topic.replication.factor: 1
+      transaction.state.log.replication.factor: 1
+      transaction.state.log.min.isr: 1
+      log.message.format.version: "2.3"
+    storage:
+      type: ephemeral
+  zookeeper:
+    replicas: 3
+    storage:
+      type: ephemeral
+  entityOperator:
+    userOperator: {}
+    topicOperator: {}
+```
+Then applying the Kafka custom resource to the cluster:
+
+```bash
+kubectl apply -f kafka.yaml
+```
+The `metrics.reporters` field can take a list of comma-separated metric reporter class names, if you need to add others.
+At runtime, Kafka will loop through this list and instantiate these metric reporters to run concurrently.
+
+At user-defined time windows, Cruise Control will read these reported metrics from the designated Kafka topic and use them for cluster monitoring and balancing.
+How to configure the connection between Cruise Control and Kafka will be described later in the post.
+
+# Accessing Zookeeper
+
+Cruise Control gets the current state and configuration of the Kafka cluster from Zookeeper.
+For estimating partition load, monitoring cluster state, and triggering partition reassignments, Cruise Control needs to be able communicate with Zookeeper directly.
+Communication between Zookeeper and all other Strimzi components is always encrypted and authenticated using TLS.
+This preventsCruise Control making a non-TLS connection  to Zookeeper.
+However, using a hack by Jakub Scholz, you can get around this requirement by creating a Kubernetes service which reuses Strimzi certificates to pipe unencrypted and unauthenticated traffic to Zookeeper. This allows Cruise Control to communicate with Zookeeper without the need of proper authorization and encryption configuration via Java system properties or Stunnel sidecars.
+
+```
+kubectl apply -f https://gist.githubusercontent.com/scholzj/6cfcf9f63f73b54eaebf60738cfdbfae/raw/068d55ac65e27779f3a5279db96bae03cea70acb/zoo-entrance.yaml
+```
+
+**WARNING**: Allowing unauthenticated access to Zookeeper exposes sensitive Kafka metadata for reading and writing by any user.
+Nefarious users could exploit this access by altering topic configurations to mess up cluster state or by changing ACL access rules to elevate client privileges for stealing data.
+
+**Note**: Kafka is in the process of migrating to using the Kafka `AdminClient` instead of Zookeeper for operations like initiating replica reassignment.
+To stay compatible with future versions of Kafka, Cruise Control will need to follow these API changes.
```

Comment:
@tombentley will these two lines suffice for mention of the AdminClient?

Reply:
TBH I think just the first is sufficient.